### PR TITLE
NIMATA!

### PR DESCRIPTION
**Pull request by OFFTKP**

### CODE REVIEW

**Comment by georgemoralis** [comment]
oti nanai einai auto den to kano merge

**Comment by georgemoralis** [comment]
pare ton rodakino na se kanei code review na matheis

**Comment by wheremyfoodat** [comment]
Συνήθως το code review γίνεται όταν ο κώδικας κάνει compile

**Comment by OFFTKP** [comment]
@wheremyfoodat

**Comment by wheremyfoodat** [comment]
Θα το φτιαξω οταν είμαι σε πισι

**Review by wheremyfoodat — APPROVED** [review]

**Comment by georgemoralis** [comment]
throw an error on this

```

if (!_Verify_ownership_levels()) {
    // only occurs for recursive mutexes (N4950 [thread.mutex.recursive]/3)
    // POSIX specifies EAGAIN in the corresponding situation:
    // https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_lock.html
    _STD _Throw_Cpp_error(_RESOURCE_UNAVAILABLE_TRY_AGAIN);
}
```

File: src/Lib/Threads.h [hunk 1]
```diff
@@ -108,9 +108,9 @@ class ConditionVariable {
 };
 
 struct ConditionVariableStructInternal {
-    ConditionVariableStructInternal() { InitializeConditionVariable(&m_cv); }
+    ConditionVariableStructInternal() = default;
     ~ConditionVariableStructInternal() = default;
-    CONDITION_VARIABLE m_cv{};
+    std::condition_variable m_cv{};
```
Comment (georgemoralis) [inline src/Lib/Threads.h:113]: std::condition_variable_any

File: src/Lib/Threads.cpp [hunk 2]
```diff
@@ -55,25 +56,46 @@ Lib::Mutex::Mutex() { m_mutex = new MutexStructInternal(); }
...
+    std::unique_lock<std::mutex> lock(mutex->m_mutex->m_cs);
+    m_cond_var->m_cv.wait(lock);
+}
 
 bool Lib::ConditionVariable::WaitCondVarFor(Mutex* mutex, u32 micros) {
-    bool ok = false;
-    ok = !(SleepConditionVariableCS(&m_cond_var->m_cv, &mutex->m_mutex->m_cs, (micros < 1000 ? 1 : micros / 1000)) == 0 &&
-           GetLastError() == ERROR_TIMEOUT);
-    return ok;
+    std::unique_lock<std::mutex> lock(mutex->m_mutex->m_cs);
```
Comment (georgemoralis) [inline src/Lib/Threads.cpp:95]: bool Lib::ConditionVariable::WaitCondVarFor(Mutex* mutex, u32 micros) {
    std::unique_lock<std::recursive_mutex> lock(mutex->m_mutex->m_cs, std::adopt_lock_t());
    bool r =m_cond_var->m_cv.wait_for(lock, std::chrono::microseconds(micros)) == std::cv_status::no_timeout;
    lock.release();
    return r;
}

File: src/Lib/Threads.cpp [hunk 3]
```diff
@@ -55,25 +56,46 @@ Lib::Mutex::Mutex() { m_mutex = new MutexStructInternal(); }
...
+    return false;
+}
 
 Lib::ConditionVariable::ConditionVariable() { m_cond_var = new ConditionVariableStructInternal(); }
 
 Lib::ConditionVariable::~ConditionVariable() { delete m_cond_var; }
 
-void Lib::ConditionVariable::WaitCondVar(Mutex* mutex) { SleepConditionVariableCS(&m_cond_var->m_cv, &mutex->m_mutex->m_cs, INFINITE); }
+void Lib::ConditionVariable::WaitCondVar(Mutex* mutex) {
+    std::unique_lock<std::mutex> lock(mutex->m_mutex->m_cs);
```
Comment (georgemoralis) [inline src/Lib/Threads.cpp:90]: void Lib::ConditionVariable::WaitCondVar(Mutex* mutex) {
    std::unique_lock<std::recursive_mutex> lock(mutex->m_mutex->m_cs, std::adopt_lock_t());
    m_cond_var->m_cv.wait(lock);
    lock.release();
}

File: src/Lib/Threads.h [hunk 4]
```diff
@@ -89,9 +88,10 @@ class Mutex {
 };
 
 struct MutexStructInternal {
-    MutexStructInternal() { InitializeCriticalSectionAndSpinCount(&m_cs, 4000); }
-    ~MutexStructInternal() { DeleteCriticalSection(&m_cs); }
-    CRITICAL_SECTION m_cs{};
+    MutexStructInternal() = default;
+    ~MutexStructInternal() = default;
+    std::mutex m_cs{};
```
Comment (georgemoralis) [inline src/Lib/Threads.h:93]: std::recursive_mutex m_cs{};

**Review by georgemoralis — COMMENTED** [review]
pariiii

**Comment by georgemoralis** [comment]
pari you can do it . We believe y

**Comment by georgemoralis** [comment]
won't fix , termination

**Pull request closed without merging**